plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,603 @@
+			g_prefix_error (error,
+					"GET_MMIO opcode [0x%x] nonzero error bit in status [0x%x]:",
+					regx->opcode, regx->status);
+			return FALSE;
+		}
+
+		/* operation valid (OV) bit should be 0'b */
+		if (regx->status & MBOX_OPVALID) {
+			g_prefix_error (error,
+					"GET_MMIO opcode [0x%x] nonzero OV bit in status [0x%x]:",
+					regx->opcode, regx->status);
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_set_mmio (FuDevice *device, guint16 mbox_reg, guchar *buf,
+			        GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    REQ_HUB_SET_MMIO, /* request */
+					    MBOX_ACCESS,      /* value */
+					    mbox_reg,	      /* index */
+					    (guchar *) buf,   /* data */
+					    4,		      /* length */
+					    NULL,	      /* actual length */
+					    MBOX_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to set mmio 0x%x:", mbox_reg);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/*
+ * Read up to 64 bytes of data from the mbox data registers to a buffer.
+ * The mailbox can hold 64 bytes of data in 16 doubleword data registers.
+ * To get data from NVM or DROM to mbox registers issue a NVM Read or DROM
+ * read operation before reading the mbox data registers.
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_read (FuDevice *device, guchar *data, guint8 length,
+				  GError **error)
+{
+	guchar *ptr = data;
+	gint	i;
+
+	if (length > 64 || length % 4) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "invalid firmware data read length %u",
+			     length);
+		return FALSE;
+	}
+	/* read 4 bytes per iteration */
+	for (i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_get_mmio (device, i, ptr, error))
+			return FALSE; /* FIXME also print a error message */
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+/*
+ * The mailbox can hold 64 bytes in 16 doubleword data registers.
+ * A NVM write operation writes data from these registers to NVM
+ * at the set offset
+ */
+static gboolean
+fu_dell_dock_usb4_mbox_data_write (FuDevice *device, const guchar *data,
+				   guint8 length, GError **error)
+{
+	guchar *ptr = (guchar *) data;
+
+	if (length > 64 || length % 4) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "invalid firmware data write length %u\n",
+			     length);
+		return FALSE;
+	}
+	/* writes 4 bytes per iteration */
+	for (gint i = 0; i < length / 4; i++) {
+		if (!fu_dell_dock_usb4_hub_set_mmio (device, i, ptr, error))
+			return FALSE;
+		ptr += 4;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_operation (FuDevice *device, guint16 opcode,
+				 guchar *metadata, GError **error)
+{
+	g_autoptr (GError) error_local = NULL;
```

Comment:
move this into the scope of the for loop and you won't need to use g_clear_error.  It will initalize each time in the loop then.